java-frontend/src/main/java/org/sonar/java/cfg/CFGLoop.java

```diff
@@ -0,0 +1,178 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.cfg;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.sonar.java.cfg.CFG.Block;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class CFGLoop {
+
+  private final CFG.Block startingBlock;
+  private final Set<CFG.Block> blocks = new HashSet<>();
+  private final Set<CFG.Block> successors = new HashSet<>();
+
+  private CFGLoop(CFG.Block block, Map<Tree, CFGLoop> container) {
+    container.put(block.terminator(), this);
+    startingBlock = block;
+    Block loopFirstBlock = block.trueBlock();
+    if (loopFirstBlock == null) {
+      // Special case where no condition is given in FOR loop: only one successor specified
+      loopFirstBlock = block.successors().iterator().next();
+    }
+    collectBlocks(loopFirstBlock, new HashSet<CFG.Block>(), container);
+    successors.addAll(block.successors());
+    successors.remove(block.falseBlock());
+    collectWaysOut(container);
+  }
+
+  public CFG.Block startingBlock() {
+    return startingBlock;
+  }
+
+  public Collection<CFG.Block> blocks() {
+    return sorted(blocks);
+  }
+
+  public Collection<CFG.Block> successors() {
+    return sorted(successors);
+  }
+
+  public boolean hasNoWayOut() {
+    return successors.isEmpty();
+  }
+
+  private boolean collectBlocks(@Nonnull CFG.Block block, Set<CFG.Block> visitedBlocks, Map<Tree, CFGLoop> container) {
+    if (blocks.contains(block)) {
+      return true;
+    }
+    if (block == null || block == startingBlock || !visitedBlocks.add(block)) {
+      return false;
+    }
+    boolean answer = returnsToStart(block, visitedBlocks, container);
+    if (answer || isBreak(block)) {
+      blocks.add(block);
+    }
+    return answer;
+  }
+
+  private boolean returnsToStart(@Nonnull CFG.Block block, Set<CFG.Block> visitedBlocks, Map<Tree, CFGLoop> container) {
+    Set<Block> localSuccessors = localSuccessors(block, container);
+    if (localSuccessors == null) {
+      return true;
+    }
+    boolean answer = false;
+    for (CFG.Block successor : localSuccessors) {
+      if (startingBlock == successor) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)